src/solvers/refinement/string_constraint_generator.cpp

```diff
@@ -0,0 +1,2236 @@
+  const function_application_exprt &f)
+{
+  string_exprt res(get_char_type());
+  string_exprt str=add_axioms_for_string_expr(args(f, 3)[0]);
+  with_exprt sarrnew(str.content(), args(f, 3)[1], args(f, 3)[2]);
+  implies_exprt lemma(
+    binary_relation_exprt(args(f, 3)[1], ID_lt, str.length()),
+    and_exprt(
+      equal_exprt(res.content(), sarrnew),
+      res.axiom_for_has_same_length_as(str)));
+  axioms.push_back(lemma);
+  return res;
+}
+
+string_exprt string_constraint_generatort::add_axioms_for_replace(
+  const function_application_exprt &f)
+{
+  string_exprt str=add_axioms_for_string_expr(args(f, 3)[0]);
+  exprt oldChar=args(f, 3)[1];
+  exprt newChar=args(f, 3)[2];
+  string_exprt res(get_char_type());
+
+  // We add axioms:
+  // a1 : |res| = |str|
+  // a2 : forall qvar, 0<=qvar<|res|,
+  //    str[qvar]=oldChar => res[qvar]=newChar
+  //    !str[qvar]=oldChar => res[qvar]=str[qvar]
+
+  axioms.push_back(res.axiom_for_has_same_length_as(str));
+
+  symbol_exprt qvar=fresh_univ_index("QA_replace");
+  implies_exprt case1(
+    equal_exprt(str[qvar], oldChar),
+    equal_exprt(res[qvar], newChar));
+  implies_exprt case2(
+    not_exprt(equal_exprt(str[qvar], oldChar)),
+    equal_exprt(res[qvar], str[qvar]));
+  string_constraintt a1(qvar, res.length(), and_exprt(case1, case2));
+  axioms.push_back(a1);
+  return res;
+}
+
+string_exprt string_constraint_generatort::add_axioms_for_delete_char_at(
+  const function_application_exprt &f)
+{
+  string_exprt str=add_axioms_for_string_expr(args(f, 2)[0]);
+  exprt index_one=refined_string_typet::index_of_int(1);
+  return add_axioms_for_delete(
+    str, args(f, 2)[1], plus_exprt(args(f, 2)[1], index_one));
+}
+
+string_exprt string_constraint_generatort::add_axioms_for_delete(
+  const string_exprt &str, const exprt & start, const exprt & end)
+{
+  assert(start.type()==refined_string_typet::index_type());
+  assert(end.type()==refined_string_typet::index_type());
+  string_exprt str1=add_axioms_for_substring(
+    str, refined_string_typet::index_zero(), start);
+  string_exprt str2=add_axioms_for_substring(str, end, str.length());
+  return add_axioms_for_concat(str1, str2);
+}
+
+string_exprt string_constraint_generatort::add_axioms_for_delete(
+  const function_application_exprt &f)
+{
+  string_exprt str=add_axioms_for_string_expr(args(f, 3)[0]);
+  return add_axioms_for_delete(str, args(f, 3)[1], args(f, 3)[2]);
+}
+
+
+string_exprt string_constraint_generatort::add_axioms_for_concat_int(
+  const function_application_exprt &f)
+{
+  string_exprt s1=add_axioms_for_string_expr(args(f, 2)[0]);
+  string_exprt s2=add_axioms_from_int(args(f, 2)[1], 10);
+  return add_axioms_for_concat(s1, s2);
+}
+
+string_exprt string_constraint_generatort::add_axioms_for_concat_long(
+  const function_application_exprt &f)
+{
+  string_exprt s1=add_axioms_for_string_expr(args(f, 2)[0]);
+  string_exprt s2=add_axioms_from_int(args(f, 2)[1], 30);
```

Comment:
Sorry I thought I corrected this, the maximal length a java Long can have is 20 characters https://docs.oracle.com/javase/7/docs/api/java/lang/Long.html#MIN_VALUE
I also corrected the maximal length for Integer (can be 11 if we count the initial `-`)
https://github.com/diffblue/cbmc/pull/374/commits/f93eb6d1dca75a4dba79ae417fab01ed35243fd6